### PR TITLE
[update TensorRT converter]

### PR DESCRIPTION
**Pull request by jjsjann123**
fixed FusedBatchNorm to support broadcast;
  remove fp16 conversion for type int const
  add Snapshot in conversion (treated as identity)

### CODE REVIEW

**Comment by jjsjann123** [comment]
Patching TensorRT converter
@aaroey @gunan

**Comment by gunan** [comment]
Could you send this to the master branch, too?
@yifeif CC, current release owner

**Comment by jjsjann123** [comment]
Sure.
Working on it right now

**Comment by yifeif** [comment]
@aaroey let me know if this is ready to be merged. Thanks.

**Comment by aaroey** [comment]
@yifeif sorry I haven't had a chance to look into this, will do probably later today.
@jjsjann123 I'm just wondering, does this PR need to be merged in 1.7? If so do you need to test everything you have done before? As @cliffwoolley mentioned that we're already good to go as-is.
Also @samikama.

**Comment by jjsjann123** [comment]
This patch is to support more slim networks. We want this to go into 1.7.
Since changes are only made to the graph converter, it will not affect the build nor Tensorflow functionalities outside the conversion.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -1880,11 +1838,53 @@ tensorflow::Status ConvertFusedBatchNorm(
   TRT_ShapedWeights mean_weights = inputs.at(3).weights();
   TRT_ShapedWeights variance_weights = inputs.at(4).weights();
   TRT_ShapedWeights dummy_power_weights(scale_weights.type_);
+  size_t scale_weights_count = scale_weights.count();
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Can you use for loops to reduce the boilerplate code from line 1836 to 1879?

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 2]
```diff
@@ -1897,12 +1897,29 @@ tensorflow::Status ConvertFusedBatchNorm(
     }
     if (scale_weights.type_ == tensorflow::DT_FLOAT) {
       for (size_t i = 0; i < nweight; ++i) {
-        float scale = (static_cast<float const*>(scale_weights.GetValues()))[i];
-        float offset =
-            (static_cast<float const*>(offset_weights.GetValues()))[i];
-        float mean = (static_cast<float const*>(mean_weights.GetValues()))[i];
-        float variance =
-            (static_cast<float const*>(variance_weights.GetValues()))[i];
+        float scale, offset, mean, variance;
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Can we use a for loop here also? Everything is from input.at(i).weights().

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 3]
```diff
@@ -1924,22 +1941,28 @@ tensorflow::Status ConvertFusedBatchNorm(
       Eigen::half* comb_off_vals = const_cast<Eigen::half*>(
           static_cast<Eigen::half const*>(combined_offset_weights.GetValues()));
       for (size_t i = 0; i < nweight; ++i) {
-        float scale(scale_vals[i]);
-        float offset(off_vals[i]);
-        float mean(mean_vals[i]);
-        float variance(variance_vals[i]);
+        int si = scale_weights_count != 1 ? i : 0;
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: For loop?

**Review by aaroey — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 4]
```diff
@@ -346,11 +346,10 @@ void ReorderCKtoKC(const TRT_ShapedWeights& iweights,
       break;
     }
     case tensorflow::DataType::DT_HALF: {
-      Reorder2(
-          {k, c}, static_cast<Eigen::half const*>(iweights.GetValues()),
-          istrides,
-          static_cast<Eigen::half*>(const_cast<void*>(oweights->GetValues())),
-          ostrides);
+      Reorder2({k, c}, static_cast<Eigen::half const*>(iweights.GetValues()),
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: This is from clang-format.
I'll revert it.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 5]
```diff
@@ -1447,62 +1446,23 @@ tensorflow::Status ConvertConst(Converter& ctx,
...
-      void* dst = static_cast<void*>(&(ctx.weight_store()->store_.back()[0]));
-      std::vector<int32> tensor_data(
-          weights_tensor.int_val().begin(),
-          weights_tensor.int_val()
-              .end());  //  make a local copy first to flatten
-                        //  doesn't have to be contigous
-      memcpy(dst, tensor_data.data(), len_tensor);  // store into weight store
-      weights = TRT_ShapedWeights(dtype, dst, scalar_shape);
-    }
+    //  we should not have converted //if (ctx.isFP16()) {
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: This fp16 conversion is converting int value to fp16.
We should only convert fp32 to fp16 (we kept that one here). int values are used for attributes for supported ops in the converter.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc (hunk 1 above)
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Originally avoiding for loop for readability. (scale, offset, mean, variance)
Will update them and put in comment.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc (hunk 2 above)
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: will do.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 6]
```diff
@@ -1875,71 +1833,105 @@ tensorflow::Status ConvertFusedBatchNorm(
...
+      float variance = batchnorm_data[3];
+      float& combined_scale_ref = const_cast<float*>(
+          static_cast<float const*>(combined_scale_weights.GetValues()))[i];
+      float& combined_offset_ref = const_cast<float*>(
+          static_cast<float const*>(combined_offset_weights.GetValues()))[i];
+      combined_scale_ref = scale / sqrtf(variance + epsilon);
+      combined_offset_ref = offset - mean * combined_scale_ref;
+    }
+  } else {
+    const Eigen::half* cast_vals_array[4];
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: I think this branch (starting from line 1908) can reuse code from line 1876 to 1897 in the `if` branch, by putting the `parameter_type` check inside the `for (size_t i..)` loop? `cast_vals_array` can be of float type as it gets assigned to `batchnorm_data` which is later used as float (line 1917)?

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 7]
```diff
@@ -1000,7 +999,7 @@ tensorflow::Status ConvertConv2DHelper(
     const std::vector<TRT_TensorOrWeights>& inputs,
     std::vector<TRT_TensorOrWeights>* outputs,
     int group  // group ==0 specifies depthwise conv
-) {
+    ) {
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Nit: please merge this line with the previous line. For the comment, either use `/**/` or put it after `) {`.

**Review by aaroey — APPROVED** [review]

**Comment by aaroey** [comment]
Thanks @jjsjann123 for the fix. PR looks good to me, will wait for @gunan for the merge.
